smbclient/shutil.py

```diff
@@ -0,0 +1,306 @@
+    links are copied.
+
+    The optional ignore argument is a callable. If given, it
+    is called with the `src` parameter, which is the directory
+    being visited by copytree(), and `names` which is the list of
+    `src` contents, as returned by os.listdir():
+
+        callable(src, names) -> ignored_names
+
+    Since copytree() is called recursively, the callable will be
+    called once for each directory that is copied. It returns a
+    list of names relative to the `src` directory that should
+    not be copied.
+
+    XXX Consider this example code rather than the ultimate tool.
+
+    """
+    names = listdir(src, **kwargs)
+    if ignore is not None:
+        ignored_names = ignore(src, names)
+    else:
+        ignored_names = set()
+
+    makedirs(dst, **kwargs)
+    errors = []
+    for name in names:
+        if name in ignored_names:
+            continue
+        srcname = join(src, name)
+        dstname = join(dst, name)
+        try:
+            if symlinks and islink(srcname, **kwargs):
+                linkto = readlink(srcname, **kwargs)
+                symlink(linkto, dstname, **kwargs)
+            elif isdir(srcname, **kwargs):
+                copytree(srcname, dstname, symlinks, ignore, **kwargs)
+            else:
+                # Will raise a SpecialFileError for unsupported file types
+                copy2(srcname, dstname, server_side_copy=server_side_copy, **kwargs)
+        # catch the Error from the recursive copytree so that we can
+        # continue with other files
+        except Error as err:
+            errors.extend(err.args[0])
+        except EnvironmentError as why:
+            errors.append((srcname, dstname, str(why)))
+    try:
+        copystat(src, dst)
+    except OSError as why:
+        if WindowsError is not None and isinstance(why, WindowsError):
+            # Copying file access times may fail on Windows
+            pass
+        else:
+            errors.append((src, dst, str(why)))
+    if errors:
+        raise Error(errors)
+
+
+def copyfile_server_side(src, dst, **kwargs):
+    """
+    Server side copy of a file
+
+    :param src: The source file.
+    :param dst: The target file.
+    :param kwargs: Common SMB Session arguments for smbclient.
+    """
+
+    norm_dst = normpath(dst)
+    norm_src = normpath(src)
+
+    src_drive = splitdrive(norm_src)[0]
+    dst_drive = splitdrive(norm_dst)[0]
+    if src_drive.lower() != dst_drive.lower():
+        raise ValueError(
+            "Server side copy can only occur on the same drive, '%s' must be the same as the dst root '%s'" % (
+                src_drive, dst_drive))
+
+    try:
+        stat(norm_dst, **kwargs)
+    except SMBOSError as err:
+        if err.errno != errno.ENOENT:
+            raise
+    else:
+        raise ValueError("Target %s already exists" % norm_dst)
+
+    with SMBRawIO(norm_src, mode='r', desired_access=FilePipePrinterAccessMask.GENERIC_READ,
```

Comment:
After rebase, the test are failing, I'll have to find out why first.

Reply:
Some of the failures were my doing sorry, I renamed some kwargs to avoid the builtin names. One I know off the top of my head is that `ioctl_request's` `buffer` kwarg is now `input_buffer`.

Reply:
Right, that was the issue!

Reply:
When changing to the suggested `open_file` I get
```
>       sid = self.raw.fd.tree_connect.session.session_id
E       AttributeError: '_io.TextIOWrapper' object has no attribute 'fd'
```

Reply:
It sounds like you need to open it in binary mode. Typically you will want to set `mode='<x>b` (the b is important) to make sure you are dealing with bytes. Internal functions should avoid using the text side unless they really need it.